repose-aggregator/src/docs/asciidoc/services/datastores.adoc

```diff
@@ -3,4 +3,293 @@
+The given key is then reduced to its numeric representation and a cluster node is looked up by performing a modulus such that (`<key-value> % <number-of-cluster-members>`).
+
+=== Key-space Encoding
+By default, the internal Repose client implementation for the distributed datastore will obscure key-space by storing only the MD5 hash value of a given key and not the key's actual value.
+This is important to note since external gets against the distributed datastore must be aware of this functionality.
+The MD5 hash is represented as a 128bit UUID.
+
+Example Key Addressing
+
+* String Key: `object-key`
+* MD5 Hash: `cd26615a30a3cdce02e3a834fed5711a`
+* UUID: `cecda330-5a61-26cd-1a71-d5fe34a8e302`
+
+If an external application makes a request for data stored by Repose components, it must first hash the key using MD5 before sending the request such that...
+
+`GET /powerapi/dist-datastore/objects/object-key`
+
+becomes
+
+`GET /powerapi/dist-datastore/objects/cecda330-5a61-26cd-1a71-d5fe34a8e302`
+
+Obscuring key-space is not a function of the distributed datastore service.
+This functionality is only present in the internally consumed java cache client.
+If an external application puts an object into the distributed datastore, the object will be stored under the value of the key given.
+
+=== Remote Management
+The repose distributed datastore component is a filter component that hosts a simple RESTful API that can be contacted to perform remote object store operations.
+These operations are defined below.
+
+`GET /powerapi/dist-datastore/objects/<object-key> HTTP/1.1`
+
+Gets a stored object from the datastore by its key.
+
+`PUT /powerapi/dist-datastore/objects/<object-key> HTTP/1.1`
+
+Puts an object into the datastore by its key.
+
+`DELETE /powerapi/dist-datastore/objects/<object-key> HTTP/1.1`
+
+Deletes an object from the datastore by its key.
+
+`PATCH /powerapi/dist-datastore/objects/<object-key> HTTP/1.1`
+
+Patches an object in the datastore by its key.
+If the object does not exist, then a new one created.
+Return the modified/new object.
+The object must be Patchable.
+
+=== Remote Fail-Over
+In the event that a node with in a datastore cluster falls off line or is unable to respond to requests, it is removed from the node's cluster membership for a period of time.
+During this time, the online node will then re-address its key-space in order to continue operation.
+After certain periods of rest, the node may attempt to introduce the damaged cluster member into its cluster membership.
+A damaged cluster member must go through several validation passes where the member is introduced back into the addressing algorithm before it can be considered online.
+In order to keep healthy nodes from attempting to route requests to the damaged node, a participating node may tell it's destination that the destination may not route the request and must store, delete or get the value locally.
```

Comment:
"store, delete or get". Not a fan of that phrasing. That seems to try and enumerate the methods that can be performed with a value, but it does not include patch. Even if you read it more generically, patch does not store "the value" locally in all cases. It modifies the existing value and stores that. I would love to see that phrase replaced with something like "handle". So "... and must handle the value locally."